ingestion-beam/src/main/java/com/mozilla/telemetry/decoder/DecryptPioneerPayloads.java

```diff
@@ -0,0 +1,87 @@
+package com.mozilla.telemetry.decoder;
+
+import com.fasterxml.jackson.databind.node.ObjectNode;
+import com.mozilla.telemetry.ingestion.core.Constant.Attribute;
+import com.mozilla.telemetry.transforms.FailureMessage;
+import com.mozilla.telemetry.transforms.PubsubConstraints;
+import com.mozilla.telemetry.util.Json;
+import com.mozilla.telemetry.util.KeyStore;
+import java.io.IOException;
+import java.security.PrivateKey;
+import java.util.Collections;
+import org.apache.beam.sdk.io.gcp.pubsub.PubsubMessage;
+import org.apache.beam.sdk.options.ValueProvider;
+import org.apache.beam.sdk.transforms.FlatMapElements;
+import org.apache.beam.sdk.transforms.PTransform;
+import org.apache.beam.sdk.transforms.ProcessFunction;
+import org.apache.beam.sdk.transforms.WithFailures;
+import org.apache.beam.sdk.transforms.WithFailures.Result;
+import org.apache.beam.sdk.values.PCollection;
+import org.apache.beam.sdk.values.TypeDescriptor;
+import org.jose4j.jwe.JsonWebEncryption;
+import org.jose4j.lang.JoseException;
+
+public class DecryptPioneerPayloads extends
+    PTransform<PCollection<PubsubMessage>, Result<PCollection<PubsubMessage>, PubsubMessage>> {
+
+  private final ValueProvider<String> metadataLocation;
+  private final ValueProvider<Boolean> kmsEnabled;
+  private transient KeyStore keyStore;
+
+  public static DecryptPioneerPayloads of(ValueProvider<String> metadataLocation,
+      ValueProvider<Boolean> kmsEnabled) {
+    return new DecryptPioneerPayloads(metadataLocation, kmsEnabled);
+  }
+
+  private DecryptPioneerPayloads(ValueProvider<String> metadataLocation,
+      ValueProvider<Boolean> kmsEnabled) {
+    this.metadataLocation = metadataLocation;
+    this.kmsEnabled = kmsEnabled;
+  }
+
+  @Override
+  public Result<PCollection<PubsubMessage>, PubsubMessage> expand(
+      PCollection<PubsubMessage> messages) {
+    return messages.apply(FlatMapElements.into(TypeDescriptor.of(PubsubMessage.class)) //
+        .via(new Fn()) //
+        .exceptionsInto(TypeDescriptor.of(PubsubMessage.class)) //
+        .exceptionsVia((WithFailures.ExceptionElement<PubsubMessage> ee) -> {
+          try {
+            throw ee.exception();
+          } catch (IOException e) {
```

Comment:
Do we also want to catch JoseException here and route those to error output?

Reply:
If a JoseException were thrown here, would it kill the entire pipeline?

Reply:
> If a JoseException were thrown here, would it kill the entire pipeline?

It would bubble up and kill the worker, causing a new worker to get spun up. If it's a transient error, that's a good thing, because the message should succeed on retry. If it's a problem with the message itself, it will eventually cause the entire pipeline to fail after 4 worker failures.